derive/src/lib.rs

```diff
@@ -20,23 +20,183 @@ extern crate alloc;
-pub fn type_def(input: TokenStream) -> TokenStream {
-	type_def::generate(input.into()).into()
+fn generate(input: TokenStream2) -> Result<TokenStream2> {
+	let mut tokens = quote! {};
+	tokens.extend(generate_type(input)?);
+	Ok(tokens)
 }
 
-#[proc_macro_derive(Metadata)]
-pub fn metadata(input: TokenStream) -> TokenStream {
-	metadata::generate(input.into()).into()
+fn generate_type(input: TokenStream2) -> Result<TokenStream2> {
+	let mut ast: DeriveInput = syn::parse2(input.clone())?;
+
+	ast.generics.type_params_mut().for_each(|p| {
+		p.bounds.push(parse_quote!(_type_metadata::Metadata));
+		p.bounds.push(parse_quote!('static));
+	});
+
+	let ident = &ast.ident;
+	let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();
+	let generic_type_ids = ast.generics.type_params().map(|ty| {
+		let ty_ident = &ty.ident;
+		quote! {
+			<#ty_ident as _type_metadata::Metadata>::meta_type()
+		}
+	});
+
+	let ast: DeriveInput = syn::parse2(input.clone())?;
+	let (type_variant, type_def) = match &ast.data {
+		Data::Struct(ref s) => (quote!(Product), generate_struct_def(s)),
+		Data::Enum(ref e) => (quote!(Sum), generate_enum_def(e)),
+		Data::Union(_) => return Err(Error::new_spanned(input, "Unions not supported")),
+	};
+
+	let has_type_impl = quote! {
+		impl #impl_generics _type_metadata::HasType for #ident #ty_generics #where_clause {
+			fn get_type() -> _type_metadata::Type {
+				_type_metadata::Type::#type_variant(
+					_type_metadata::TypeComposite::new(
+						stringify!(#ident),
+						_type_metadata::Namespace::from_module_path(module_path!())
+							.expect("namespace from module path cannot fail"),
+						__core::vec![ #( #generic_type_ids ),* ],
+						#type_def.into(),
+					)
+				)
+			}
+		}
+	};
+
+	Ok(impl_wrapper::wrap(ident, "HAS_TYPE", has_type_impl))
+}
+
+type FieldsList = Punctuated<Field, Comma>;
+
+fn generate_fields_def(fields: &FieldsList) -> TokenStream2 {
+	let fields_def = fields.iter().map(|f| {
+		let (ty, ident) = (&f.ty, &f.ident);
+		let meta_type = quote! {
+			<#ty as _type_metadata::Metadata>::meta_type()
+		};
+		if let Some(i) = ident {
+			quote! {
+				_type_metadata::NamedField::new(stringify!(#i), #meta_type)
+			}
+		} else {
+			quote! {
+				_type_metadata::UnnamedField::new(#meta_type)
+			}
+		}
+	});
+	quote! { __core::vec![#( #fields_def, )*] }
+}
+
+fn generate_struct_def(data_struct: &DataStruct) -> TokenStream2 {
+	match data_struct.fields {
+		Fields::Named(ref fs) => {
+			let fields = generate_fields_def(&fs.named);
+			quote! {
+				_type_metadata::TypeProductStruct::new(#fields)
+			}
+		}
+		Fields::Unnamed(ref fs) => {
+			let fields = generate_fields_def(&fs.unnamed);
+			quote! {
+				_type_metadata::TypeProductTupleStruct::new(#fields)
+			}
+		}
+		Fields::Unit => quote! {
+			_type_metadata::TypeProductTupleStruct::unit()
+		},
+	}
+}
+
+type VariantList = Punctuated<Variant, Comma>;
+
+fn generate_c_like_enum_def(variants: &VariantList) -> TokenStream2 {
```

Comment:
This `ClikeEnumVariant` should be unified to just be a variant with the mentioned properties. A build pattern might safely construct those different properties.